coderd/database/queries/notifications.sql

```diff
@@ -0,0 +1,139 @@
+                           AND (
+                             -- exclude all messages which have exceeded the max attempts; these will be purged later
+                             nm.attempt_count IS NULL OR nm.attempt_count < sqlc.arg('max_attempt_count')::int
+                             )
+                           -- if set, do not retry until we've exceeded the wait time
+                           AND (
+                             CASE
+                                 WHEN nm.next_retry_after IS NOT NULL THEN nm.next_retry_after < NOW()
+                                 ELSE true
+                                 END
+                             )
+                         ORDER BY nm.created_at ASC
+                                  -- Ensure that multiple concurrent readers cannot retrieve the same rows
+                             FOR UPDATE OF nm
+                                 SKIP LOCKED
+                         LIMIT sqlc.arg('count'))
+            RETURNING id)
+SELECT
+    -- message
+    nm.id,
+    nm.payload,
+    nm.method,
+    nm.created_by,
+    -- template
+    nt.title_template,
+    nt.body_template
+FROM acquired
+         JOIN notification_messages nm ON acquired.id = nm.id
+         JOIN notification_templates nt ON nm.notification_template_id = nt.id;
+
+-- name: BulkMarkNotificationMessagesFailed :execrows
+WITH new_values AS (SELECT UNNEST(@ids::uuid[])                             AS id,
+                           UNNEST(@failed_ats::timestamptz[])               AS failed_at,
+                           UNNEST(@statuses::notification_message_status[]) AS status,
+                           UNNEST(@status_reasons::text[])                  AS status_reason)
+UPDATE notification_messages
+SET updated_at       = subquery.failed_at,
+    attempt_count    = attempt_count + 1,
+    status           = CASE
+                           WHEN attempt_count + 1 < @max_attempts::int THEN subquery.status
+                           ELSE 'permanent_failure'::notification_message_status END,
+    status_reason    = subquery.status_reason,
+    leased_until     = NULL,
+    next_retry_after = CASE
+                           WHEN (attempt_count + 1 < @max_attempts::int)
+                               THEN NOW() + CONCAT(@retry_interval::int, ' seconds')::interval END
+FROM (SELECT id, status, status_reason, failed_at
+      FROM new_values) AS subquery
+WHERE notification_messages.id = subquery.id;
+
+-- name: BulkMarkNotificationMessagesSent :execrows
+WITH new_values AS (SELECT UNNEST(@ids::uuid[])             AS id,
+                           UNNEST(@sent_ats::timestamptz[]) AS sent_at)
+UPDATE notification_messages
+SET updated_at       = subquery.sent_at,
+    attempt_count    = attempt_count + 1,
+    status           = 'sent'::notification_message_status,
+    status_reason    = NULL,
+    leased_until     = NULL,
+    next_retry_after = NULL
+FROM (SELECT id, sent_at
+      FROM new_values) AS subquery
```

Comment:
do we really need both the `WITH` clause and this clause remapping `new_values` to `subquery`?  Is it not supported to just have

```
FROM (SELECT UNNEST(@ids::uuid[])             AS id,
             UNNEST(@sent_ats::timestamptz[]) AS sent_at) as new_values
```